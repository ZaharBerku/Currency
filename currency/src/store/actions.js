import * as actions from "./storeComponents/actions";

export default actions;