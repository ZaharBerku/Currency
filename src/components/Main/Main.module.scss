@import "../../style/mixins.scss";
@import "../../style/variables.scss";
@import "../../style/function.scss";

.main{
    background-color:  #F0F0F0;
    padding-top:  adaptiv-indentation(30,102,320);
    padding-bottom: 302px;
    &TitleText{
        font-weight: 600;
        @include adaptiv-font(32, 26);
        line-height: 54px;
        color: $primary-color;
    };
    &Container{
        @include flex-container;
        gap: adaptiv-indentation(30,47,993);
    };
    &Currency{
        @include flex-container;
        background: #FFFFFF;
        box-shadow: 4px 4px 8px rgba(70, 21, 167, 0.12);
        border-radius: 21px;
        padding: adaptiv-indentation(20,40,993);
        column-gap: adaptiv-indentation(15,40,993);
    };
    &InputAndSelect{
        display: flex;
        flex-direction: column;
        gap: 23px;   
    };
    &Input{
        box-shadow: 0px 4px 4px #E5E5E5;
        border-radius: 13px;
        padding:7px adaptiv-indentation(38,50,993);
        border: 2px solid transparent;
        &:focus-visible,
        &:focus{
            outline: 0;
            border: 2px solid $primary-color;
        };
        &:hover{
            box-shadow: 0px 0px 4px ;
        };
        &:active{
            box-shadow: inset 0px 0px 4px;
        };
    };
    &Img{
        max-height: adaptiv-indentation(220,300,993);
    };
    &BtnReverse{
        background: #FFFFFF;
        box-shadow: 0px 4px 4px #E5E5E5;
        border-radius: 13px;
        border:none;
        @include flex-container;
        padding: adaptiv-indentation(8,10,993);
        &:hover{
            box-shadow: 0px 0px 4px ;
        };
        &:active{
            box-shadow: inset 0px 0px 4px;
        };
    };
    @include phone{
        background-color:  #F0F0F0;
        padding-bottom: 302px;
        &TitleText{
            font-weight: 600;
            @include adaptiv-font(32, 26);
            line-height: 54px;
            color: $primary-color;
        };
        &Container{
            @include flex-container;
            gap: 87px;
        };
        &Currency{
            @include flex-container;
            flex-direction: column;
            gap: 20px;
        };
        &InputAndSelect{
            display: flex;
            flex-direction: column;
            gap: 23px;   
        };
        &Input{
            box-shadow: 0px 4px 4px #E5E5E5;
            border-radius: 13px;
            padding:7px 38px;
            border: 2px solid transparent;
            &:focus-visible,
            &:focus{
                outline: 0;
                border: 2px solid $primary-color;
            };
            &:hover{
                box-shadow: 0px 0px 4px ;
            };
            &:active{
                box-shadow: inset 0px 0px 4px;
            };
        };
        &Img{
            display: none
        };
        &BtnReverse{
            background: #FFFFFF;
            box-shadow: 0px 4px 4px #E5E5E5;
            border-radius: 13px;
            border:none;
            @include flex-container;
            &:hover{
                box-shadow: 0px 0px 4px ;
            };
            &:active{
                box-shadow: inset 0px 0px 4px;
            };
        }; 
    };
    @include tablet{
        &Img{
            display: none;
        };
        &Currency{
            @include flex-container;
            flex-direction: column;
            gap: 20px;
        };
    };
};
