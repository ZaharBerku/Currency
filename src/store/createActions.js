export {
      //SELECT
  getOptions,
  filterOptions,
  reverseValueSelect,
  getCcy,
  getBaseCcy,
  
  //INPUT
  changeInputValue,
  getValueInput,
  isDisableInput,
  reverseValueInputs,

  //CURRENCY
  getDataCurrency,

  //HEADER 
  getDataHeader
} from "./storeComponents/createActions";

