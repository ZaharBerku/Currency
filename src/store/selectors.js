import * as selectors from "./storeComponents/select";

export { selectors };