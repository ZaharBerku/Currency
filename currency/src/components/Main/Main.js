import React from "react";
import styles from "./Main.module.scss";
import Input from "../Input";
import Button from "../Button";
import Select from "react-select";
import { colourStyles } from "../../data";
import createActions  from "../../store/createActions";
import actions from "../../store/actions";
import { useDispatch, useSelector } from "react-redux";
import bigBoss from "../../assets/images/big-boss.png";
import GlobalSvg from "../../assets/icons/global/GlobalSvg";




const Main = () => {
    const dispatch = useDispatch();
    const { input, select, currency:{data} } = useSelector(state => state);
    const {valueInputFirst, valueInputSecond, isDisable} = input;
    const { optionsSelectFirst, optionsSelectSecond, ccy, base_ccy } = select;
    
    const { 
        getBaseCcy, 
        getCcy, 
        getValueInput, 
        changeInputValue, 
        filterOptions, 
        reverseValueInputs,
        reverseValueSelect } = createActions;


    const { FILTER_OPTIONS_SELECT_FIRST, FILTER_OPTIONS_SELECT_SECOND } = actions;


    return (
        <main className={styles.main}>
            <div className={styles.main__container}>
                <div>
                    <h1 className={styles.main__title_text}>Instant currency exchange</h1>
                    <div className={styles.main__currency}>
                        <div className={styles.main__input_and_select}>
                            <Select
                                value={{ value: ccy, label: ccy }}
                                styles={colourStyles}
                                options={optionsSelectFirst}
                                onChange={({ value }) => {
                                    dispatch(getCcy(value));
                                    dispatch(changeInputValue({
                                        currency:data,
                                        nameFirst: "valueInputFirst",
                                        nameSecond: "valueInputSecond",
                                        ccy,
                                        base_ccy
                                    }));
                                    dispatch(filterOptions(FILTER_OPTIONS_SELECT_SECOND, value));
                                }} />

                            <Input
                                isDisable={isDisable}
                                value={valueInputFirst}
                                type={"number"}
                                handelInput={({ target }) => {
                                    dispatch(getValueInput({ name: "valueInputFirst", value: target.value }));
                                    dispatch(changeInputValue(
                                        {
                                            currency:data,
                                            nameFirst: "valueInputFirst",
                                            nameSecond: "valueInputSecond",
                                            ccy,
                                            base_ccy
                                        }
                                    ));  
                                }}
                                additionalClassNames={styles.main__input}
                                placeholder={"0,000"} />
                        </div>


                        <Button 
                        additionalClassNames={styles.main__btn_reverse}
                        handelClick={()=> {
                            dispatch(reverseValueInputs());
                            dispatch(reverseValueSelect());
                            dispatch(filterOptions(FILTER_OPTIONS_SELECT_FIRST, ccy));
                            dispatch(filterOptions(FILTER_OPTIONS_SELECT_SECOND, base_ccy));
                        }}>
                            <GlobalSvg type={"arrowReverse"}/>
                        </Button>


                        <div className={styles.main__input_and_select}>
                            <Select
                                value={{ value: base_ccy, label: base_ccy }}
                                styles={colourStyles}
                                options={optionsSelectSecond}
                                onChange={({ value }) => {
                                    dispatch(getBaseCcy(value));
                                    dispatch(changeInputValue({
                                        currency:data,
                                        nameFirst: "valueInputSecond",
                                        nameSecond: "valueInputFirst",
                                        ccy,
                                        base_ccy:value
                                    }));
                                    dispatch(filterOptions(FILTER_OPTIONS_SELECT_FIRST, value));
                                }} />


                            <Input
                                isDisable={isDisable}
                                value={valueInputSecond}
                                type={"number"}
                                handelInput={({ target }) => {
                                    dispatch(getValueInput({ name: "valueInputSecond", value: target.value }));
                                    dispatch(changeInputValue(
                                        {
                                            currency:data,
                                            nameFirst: "valueInputSecond",
                                            nameSecond: "valueInputFirst",
                                            ccy,
                                            base_ccy
                                        }
                                    ));
                                }}
                                additionalClassNames={styles.main__input}
                                placeholder={"0,000"} />
                        </div>
                    </div>
                </div>
                <img className={styles.main__img} src={bigBoss} alt="big-boss" />
            </div>
        </main>
    );
};


export default Main;