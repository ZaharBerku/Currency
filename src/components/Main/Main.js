import React from "react";
import { useSelector } from "react-redux";
import Select from "react-select";
import styles from "./Main.module.scss";
import { Input } from "../Input";
import { Button } from "../Button";
import { colourStyles } from "../../data";
import { selectors } from "../../store/selectors";
import { FILTER_OPTIONS_SELECT_FIRST, FILTER_OPTIONS_SELECT_SECOND } from "../../store/actions";
import bigBoss from "../../assets/images/big-boss.png";
import { GlobalSvg } from "../../assets/icons/global/GlobalSvg";
import { useCreateActions } from "../../hooks/useCreateActions";
import { createActions }  from "../../store/createActions";



const Main = () => {
    const { data } = useSelector(selectors.currency);
    const {valueInputFirst, valueInputSecond, isDisable} = useSelector(selectors.input);
    const { optionsSelectFirst, optionsSelectSecond, ccy, base_ccy } = useSelector(selectors.select);
    const { getValueInput, getCcy, changeInputValue, filterOptions, reverseValueInputs, reverseValueSelect, getBaseCcy} = useCreateActions(createActions);
    
    return (
        <main className={styles.main}>
            <div className={styles.mainContainer}>
                <div>
                    <h1 className={styles.mainTitleText}>Instant currency exchange</h1>
                    <div className={styles.mainCurrency}>
                        <div className={styles.mainInputAndSelect}>
                            <Select
                                value={{ value: ccy, label: ccy }}
                                styles={colourStyles}
                                options={optionsSelectFirst}
                                onChange={({ value }) => {
                                    getCcy(value);
                                    changeInputValue({
                                        currency:data,
                                        nameFirst: "valueInputFirst",
                                        nameSecond: "valueInputSecond",
                                        ccy,
                                        base_ccy
                                    });
                                    filterOptions(FILTER_OPTIONS_SELECT_SECOND, value);
                                }} />

                            <Input
                                isDisable={isDisable}
                                value={valueInputFirst}
                                type={"number"}
                                handelInput={({ target }) => {
                                    getValueInput({ name: "valueInputFirst", value: target.value });
                                    changeInputValue(
                                        {
                                            currency:data,
                                            nameFirst: "valueInputFirst",
                                            nameSecond: "valueInputSecond",
                                            ccy,
                                            base_ccy
                                        }
                                    );  
                                }}
                                additionalClassNames={styles.mainInput}
                                placeholder={"0,000"} />
                        </div>


                        <Button 
                        additionalClassNames={styles.mainBtnReverse}
                        handelClick={()=> {
                            reverseValueInputs();
                            reverseValueSelect();
                            filterOptions(FILTER_OPTIONS_SELECT_FIRST, ccy);
                            filterOptions(FILTER_OPTIONS_SELECT_SECOND, base_ccy);
                        }}>
                            <GlobalSvg type={"arrowReverse"}/>
                        </Button>


                        <div className={styles.mainInputAndSelect}>
                            <Select
                                value={{ value: base_ccy, label: base_ccy }}
                                styles={colourStyles}
                                options={optionsSelectSecond}
                                onChange={({ value }) => {
                                    getBaseCcy(value);
                                    changeInputValue({
                                        currency:data,
                                        nameFirst: "valueInputSecond",
                                        nameSecond: "valueInputFirst",
                                        ccy,
                                        base_ccy:value
                                    });
                                    filterOptions(FILTER_OPTIONS_SELECT_FIRST, value);
                                }} />


                            <Input
                                isDisable={isDisable}
                                value={valueInputSecond}
                                type={"number"}
                                handelInput={({ target }) => {
                                    getValueInput({ name: "valueInputSecond", value: target.value });
                                    changeInputValue(
                                        {
                                            currency:data,
                                            nameFirst: "valueInputSecond",
                                            nameSecond: "valueInputFirst",
                                            ccy,
                                            base_ccy
                                        }
                                    );
                                }}
                                additionalClassNames={styles.mainInput}
                                placeholder={"0,000"} />
                        </div>
                    </div>
                </div>
                <img className={styles.mainImg} src={bigBoss} alt="big-boss" />
            </div>
        </main>
    );
};


export { Main };