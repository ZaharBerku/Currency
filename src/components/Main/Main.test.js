/* eslint-disable no-undef */
import React from "react";
import { Main } from "./Main";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../store/store";

const Component = () => {
    return(
        <Provider store={store}>
            <Main/>
        </Provider>
    );
};

describe("Main snapshot render", () => {
    test("should Main render", () => {
        const { asFragment } = render(<Component/>);
        expect(asFragment()).toMatchSnapshot();
    });
});