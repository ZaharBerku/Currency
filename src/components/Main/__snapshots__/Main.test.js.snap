// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main snapshot render should Main render 1`] = `
<DocumentFragment>
  <main
    class="main"
  >
    <div
      class="mainContainer"
    >
      <div>
        <h1
          class="mainTitleText"
        >
          Instant currency exchange
        </h1>
        <div
          class="mainCurrency"
        >
          <div
            class="mainInputAndSelect"
          >
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-2-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class=" css-wqq7bk-control"
              >
                <div
                  class=" css-319lph-ValueContainer"
                >
                  <div
                    class=" css-qc6sy-singleValue"
                  >
                    EUR
                  </div>
                  <div
                    class=" css-6j8wv5-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-2-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <input
              class="mainInput"
              data-testid="input"
              placeholder="0,000"
              type="number"
              value="0"
            />
          </div>
          <button
            class="mainBtnReverse"
            data-testid="button"
          >
            <svg
              fill="none"
              height="29"
              viewBox="0 0 27 29"
              width="27"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.75 27.4167L1.875 23.5417L5.75 19.6667M21.25 1.58333L25.125 5.45833L21.25 9.33333V1.58333Z"
                stroke="#752AFF"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2.58333"
              />
              <path
                d="M25.125 5.45829H10.9167C8.51867 5.45829 6.21888 6.41089 4.52324 8.10653C2.8276 9.80217 1.875 12.102 1.875 14.5M1.875 23.5416H16.0833C18.4813 23.5416 20.7811 22.589 22.4768 20.8934C24.1724 19.1977 25.125 16.898 25.125 14.5"
                stroke="#752AFF"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2.58333"
              />
            </svg>
          </button>
          <div
            class="mainInputAndSelect"
          >
            <div
              class=" css-b62m3t-container"
            >
              <span
                class="css-1f43avz-a11yText-A11yText"
                id="react-select-3-live-region"
              />
              <span
                aria-atomic="false"
                aria-live="polite"
                aria-relevant="additions text"
                class="css-1f43avz-a11yText-A11yText"
              />
              <div
                class=" css-wqq7bk-control"
              >
                <div
                  class=" css-319lph-ValueContainer"
                >
                  <div
                    class=" css-qc6sy-singleValue"
                  >
                    UAH
                  </div>
                  <div
                    class=" css-6j8wv5-Input"
                    data-value=""
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-haspopup="true"
                      autocapitalize="none"
                      autocomplete="off"
                      autocorrect="off"
                      class=""
                      id="react-select-3-input"
                      role="combobox"
                      spellcheck="false"
                      style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                      tabindex="0"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
                <div
                  class=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    class=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    class=" css-tlfecz-indicatorContainer"
                  >
                    <svg
                      aria-hidden="true"
                      class="css-tj5bde-Svg"
                      focusable="false"
                      height="20"
                      viewBox="0 0 20 20"
                      width="20"
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <input
              class="mainInput"
              data-testid="input"
              placeholder="0,000"
              type="number"
              value="0"
            />
          </div>
        </div>
      </div>
      <img
        alt="big-boss"
        class="mainImg"
        src="big-boss.png"
      />
    </div>
  </main>
</DocumentFragment>
`;
